File: go/problems/pg42885/solution.go
// Package pg42885 - [PG] 42885 - 구명보트
// https://programmers.co.kr/learn/courses/30/lessons/42885
// 난이도: lv2
// 태그: greedy
//
// 시간복잡도: O(NLogN)
// 공간복잡도: O(1)
package pg42885

import "sort"

func Solution(people []int, limit int) int {

	sort.Sort(sort.Reverse(sort.IntSlice(people)))
	left, right := 0, len(people)-1

	answer := 0
	for left <= right {
		leftPerson := people[left]

		if leftPerson+people[right] <= limit {
			right--
		}

		left++
		answer++
	}
	return answer
}

File: java/src/main/java/problems/pg42885/Solution.java
package problems.pg42885;

import java.util.Arrays;

/**
 * [PG] 42885 - 구명보트 https://programmers.co.kr/learn/courses/30/lessons/42885 난이도: lv2 태그: greedy
 *
 * <p>시간복잡도: O(NlogN) 공간복잡도: O(1)
 */
class Solution {
  public int solution(int[] people, int limit) {
    int answer = 0;
    Arrays.sort(people);
    int left = 0;
    int n = people.length;
    int right = n - 1;

    while (left <= right) {

      final int person = people[right];

      if (person + people[left] <= limit) {
        left++;
      }
      answer++;
      right--;
    }
    return answer;
  }
}
